## Adding more features to our dataset

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
import os 

In [42]:
raw_dataPath = "../Raw Data/New Data"
files_list = os.listdir(raw_dataPath)

### List of all new features and dataset added

In [72]:
for idx,file in enumerate(files_list):
    print(idx, file)

0 Median Sales Price for New Houses Sold.csv
1 .DS_Store
2 Real Private Residential Fixed Investment.csv
3 Unemployment Rate - Black or African American.csv
4 Total Construction Spending- Residential.csv
5 Personal Income.csv
6 AverageHourlySalary-Private.csv
7 Job Openings- Total Nonfarm.csv
8 Unemployment Rate.csv
9 Unemployment Rate - White.csv
10 Monthly Supply of Houses in the United States.csv
11 Homeownership Rate for the United States.csv
12 Real Personal Income.csv
13 Insured Unemployment Rate.csv
14 Median Sales Price for New Houses Sold in the United States.csv
15 Number Unemployed for 27 Weeks & Over.csv
16 EMRATIO.csv


### New Features related to employment and wages

In [55]:
unemployment_rate_color = pd.read_csv(f"{raw_dataPath}/{files_list[3]}", parse_dates=[0])
unemployment_rate_white = pd.read_csv(f"{raw_dataPath}/{files_list[9]}", parse_dates=[0])
unemployment_rate = pd.read_csv(f"{raw_dataPath}/{files_list[8]}", parse_dates=[0])
insured_unemployment_rate = pd.read_csv(f"{raw_dataPath}/{files_list[13]}", parse_dates=[0])
long_unemployed = pd.read_csv(f"{raw_dataPath}/{files_list[15]}", parse_dates=[0])
emratio = pd.read_csv(f"{raw_dataPath}/{files_list[16]}", parse_dates=[0])

In [78]:
personal_income = pd.read_csv(f"{raw_dataPath}/{files_list[5]}", parse_dates=[0])
avg_hr_salary = pd.read_csv(f"{raw_dataPath}/{files_list[6]}", parse_dates=[0])
job_openings = pd.read_csv(f"{raw_dataPath}/{files_list[7]}", parse_dates=[0])
real_personal_income = pd.read_csv(f"{raw_dataPath}/{files_list[12]}", parse_dates=[0])

In [60]:
files_list[8]

'Unemployment Rate.csv'

### `chech_df` for verifying the seasonality and date range of the different datatables

In [61]:
def check_df(df):
    print(df.shape)
    print(df.columns)
    print(df["DATE"].max(), df["DATE"].min())

In [62]:
check_df(unemployment_rate)

(252, 2)
Index(['DATE', 'UNRATE'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [63]:
check_df(unemployment_rate_color)

(252, 2)
Index(['DATE', 'LNS14000006'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [64]:
check_df(unemployment_rate_white)

(252, 2)
Index(['DATE', 'LNS14000003'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [65]:
check_df(insured_unemployment_rate)

(252, 2)
Index(['DATE', 'IURNSA'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [66]:
check_df(long_unemployed)

(252, 2)
Index(['DATE', 'UEMP27OV'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [67]:
long_unemployed

DATE  UEMP27OV
0   2000-01-01       721
1   2000-02-01       629
2   2000-03-01       646
3   2000-04-01       599
4   2000-05-01       643
..         ...       ...
247 2020-08-01      1593
248 2020-09-01      2405
249 2020-10-01      3534
250 2020-11-01      3929
251 2020-12-01      3956

[252 rows x 2 columns]

May be we will not be using `long_unemployed` as it gives numbers (Thousand of Persons) whereas all other paramenter are some form of rates. And `long_unemployed` will already be considered in `unemployment_rate`.

In [68]:
check_df(emratio)

(252, 2)
Index(['DATE', 'EMRATIO'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [74]:
check_df(personal_income)

(252, 2)
Index(['DATE', 'PI'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [75]:
check_df(real_personal_income)

(252, 2)
Index(['DATE', 'RPI'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [79]:
check_df(job_openings)

(241, 2)
Index(['DATE', 'JTSJOL'], dtype='object')
2020-12-01 00:00:00 2000-12-01 00:00:00


There are only 241 rows, but other monthly data has 252 rows. We need to check what and why of missing values?

Data is starting from December 2000 whereas other dataframes have start date of Jan 2000. Thats why 11 rows are missing. Data before December was not available.


In [77]:
check_df(avg_hr_salary)

(178, 2)
Index(['DATE', 'CES0500000003'], dtype='object')
2020-12-01 00:00:00 2006-03-01 00:00:00


Average_Hourly_Salary-Private has $63$ out of $241$ columns missing i.e. $26$% of data. 

We could look to impute the missing values in above two dataframes `job_openings` & `avg_hr_salary` later on.

### Merge all above data about employment, job and salary into a common dataframe for easier utilization.

### Before merging we need to rename the columns for easy understanding.

In [93]:
def rename_column(df, name):
    return df.rename(columns={df.columns[1]:name[:-4]})

In [96]:
avg_hr_salary = rename_column(avg_hr_salary, files_list[6])

In [98]:
unemployment_rate = rename_column(unemployment_rate, files_list[8])

In [102]:
unemployment_rate_color = rename_column(unemployment_rate_color, files_list[3])
unemployment_rate_color

DATE  Unemployment Rate - Black or African American
0   2000-01-01                                            8.2
1   2000-02-01                                            8.1
2   2000-03-01                                            7.4
3   2000-04-01                                            7.0
4   2000-05-01                                            7.7
..         ...                                            ...
247 2020-08-01                                           12.8
248 2020-09-01                                           12.0
249 2020-10-01                                           10.8
250 2020-11-01                                           10.3
251 2020-12-01                                            9.9

[252 rows x 2 columns]

In [103]:
unemployment_rate_white = rename_column(unemployment_rate_white, files_list[9])

In [106]:
insured_unemployment_rate = rename_column(insured_unemployment_rate, files_list[13])


In [107]:
long_unemployed = rename_column(long_unemployed, files_list[15])

In [110]:
emratio = rename_column(emratio, files_list[16])

In [111]:
files_list[5], files_list[12]

('Personal Income.csv', 'Real Personal Income.csv')

In [112]:
personal_income = rename_column(personal_income, files_list[5])
real_personal_income = rename_column(real_personal_income, files_list[12])

In [113]:
job_openings = rename_column(job_openings, files_list[7])

In [129]:
merged_Df = unemployment_rate.merge(unemployment_rate_color, how="outer", on="DATE")

In [131]:
merged_Df = merged_Df.merge(unemployment_rate_white, how="outer", on="DATE")

In [133]:
merged_Df = merged_Df.merge(insured_unemployment_rate, how="outer", on="DATE")
merged_Df = merged_Df.merge(long_unemployed, how="outer", on="DATE")

In [135]:
merged_Df.shape

(252, 6)

In [136]:
merged_Df = merged_Df.merge(emratio, on="DATE", how="outer")
merged_Df = merged_Df.merge(personal_income, on="DATE", how="outer")
merged_Df = merged_Df.merge(real_personal_income, on="DATE", how="outer")

In [137]:
merged_Df.shape

(252, 9)

In [141]:
merged_Df = merged_Df.merge(job_openings, on="DATE", how="outer")
merged_Df = merged_Df.merge(avg_hr_salary, on="DATE", how="outer")

In [142]:
merged_Df.shape

(252, 11)

In [143]:
merged_Df.columns

Index(['DATE', 'Unemployment Rate',
       'Unemployment Rate - Black or African American',
       'Unemployment Rate - White', 'Insured Unemployment Rate',
       'Number Unemployed for 27 Weeks & Over', 'EMRATIO', 'Personal Income',
       'Real Personal Income', 'Job Openings- Total Nonfarm',
       'AverageHourlySalary-Private'],
      dtype='object')

In [144]:
merged_Df.to_csv("../Dataset/Employment_Related.csv")

Killing the kernel and reload the merged file.

In [3]:
import numpy as np
import pandas as pd
import matplotlib as plt
import os

### Remaining datatables and merging all the different datasets old and new

In [30]:
path = "../Dataset/Employment_Related.csv"

In [31]:
df = pd.read_csv(path, parse_dates=[0])

In [16]:
len(df.columns)

11

In [38]:
dataset = pd.read_csv("Dataset.csv", parse_dates=[0])

In [39]:
dataset.columns

Index(['DATE', 'MORTGAGE15US_Monthly Change', 'MORTGAGE30US_Monthly Change',
       'MORTGAGE5US_Monthly Change', 'Rent Index Mon_Chg', 'NASDAQCOM Mon_Chg',
       'WILL5000 Per_Mon_Chg', 'House Completed', 'House Under Construction',
       '5yr_Maturity Per_Mon_Chg', '10yr_Maturity Per_Mon_Chg',
       '30yr_Maturity Per_Mon_Chg', 'CPI Mon_Chg',
       'House Index Monthly Change', 'Year', 'Month'],
      dtype='object')

In [20]:
dataset.shape

(252, 16)

In [21]:
df.shape

(252, 11)

In [2]:
files_path = "../Raw Data/New Data/"

In [5]:
files = os.listdir(files_path)

In [6]:
for idx, file in enumerate(files):
    print(idx, file)

0 Median Sales Price for New Houses Sold.csv
1 .DS_Store
2 Real Private Residential Fixed Investment.csv
3 Unemployment Rate - Black or African American.csv
4 Total Construction Spending- Residential.csv
5 Personal Income.csv
6 AverageHourlySalary-Private.csv
7 Job Openings- Total Nonfarm.csv
8 Unemployment Rate.csv
9 Unemployment Rate - White.csv
10 Monthly Supply of Houses in the United States.csv
11 Homeownership Rate for the United States.csv
12 Real Personal Income.csv
13 Insured Unemployment Rate.csv
14 Median Sales Price for New Houses Sold in the United States.csv
15 Number Unemployed for 27 Weeks & Over.csv
16 EMRATIO.csv


In [10]:
new_house_median_price = pd.read_csv(f"{files_path}/{files[0]}", parse_dates=[0])
pvt_resd_inv = pd.read_csv(f"{files_path}/{files[2]}", parse_dates=[0])
cons_spend_resd = pd.read_csv(f"{files_path}/{files[4]}", parse_dates=[0])
month_supply = pd.read_csv(f"{files_path}/{files[10]}", parse_dates= [0])
ownership_rate = pd.read_csv(f"{files_path}/{files[11]}", parse_dates=[0])


In [12]:
def check_df(df):
    print(df.shape)
    print(df.columns)
    print(df["DATE"].max(), df["DATE"].min())

In [14]:
check_df(new_house_median_price)

(252, 2)
Index(['DATE', 'MSPNHSUS'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [15]:
check_df(pvt_resd_inv)

(76, 2)
Index(['DATE', 'PRFIC1'], dtype='object')
2020-10-01 00:00:00 2002-01-01 00:00:00


Real Private Residential Fixed Investment `pvt_resd_inv` is quaterly data, starting from 2002. Either we will be dropping this or imputing. 

In [16]:
check_df(cons_spend_resd)

(228, 2)
Index(['DATE', 'TLRESCONS'], dtype='object')
2020-12-01 00:00:00 2002-01-01 00:00:00


Total Construction Spending- Residential has only 228 rows, starting from 2002. Again missing values need imputation.

In [17]:
check_df(month_supply)

(252, 2)
Index(['DATE', 'MSACSR'], dtype='object')
2020-12-01 00:00:00 2000-01-01 00:00:00


In [18]:
check_df(ownership_rate)

(84, 2)
Index(['DATE', 'RSAHORUSQ156S'], dtype='object')
2020-10-01 00:00:00 2000-01-01 00:00:00


Ownership Rate is also quaterly data. 

### Renaming the column before merging them into a common dataframe.

In [19]:
def rename_column(df, name):
    return df.rename(columns={df.columns[1]:name[:-4]})

In [21]:
new_house_median_price = rename_column(new_house_median_price, files[0])
pvt_resd_inv = rename_column(pvt_resd_inv, files[2])
cons_spend_resd = rename_column(cons_spend_resd, files[4])
month_supply = rename_column(month_supply, files[10])
ownership_rate = rename_column(ownership_rate, files[11])

In [25]:
ownership_cons_df = new_house_median_price.merge(pvt_resd_inv, on="DATE", how="outer").merge(cons_spend_resd, on="DATE", how="outer")

In [27]:
ownership_cons_df = ownership_cons_df.merge(month_supply, on="DATE", how="outer")
ownership_cons_df = ownership_cons_df.merge(ownership_rate, on="DATE", how="outer")

In [28]:
ownership_cons_df.columns

Index(['DATE', 'Median Sales Price for New Houses Sold',
       'Real Private Residential Fixed Investment',
       'Total Construction Spending- Residential',
       'Monthly Supply of Houses in the United States',
       'Homeownership Rate for the United States'],
      dtype='object')

In [29]:
ownership_cons_df.to_csv("../Dataset/Ownership_Construction.csv", index=False)

In [32]:
df.columns

Index(['DATE', 'Unemployment Rate',
       'Unemployment Rate - Black or African American',
       'Unemployment Rate - White', 'Insured Unemployment Rate',
       'Number Unemployed for 27 Weeks & Over', 'EMRATIO', 'Personal Income',
       'Real Personal Income', 'Job Openings- Total Nonfarm',
       'AverageHourlySalary-Private'],
      dtype='object')

In [33]:
merged_df = df.merge(ownership_cons_df, on="DATE", how="outer")

In [40]:
merged_df = dataset.merge(merged_df, on="DATE", how="outer")

In [43]:
merged_df.columns

Index(['DATE', 'MORTGAGE15US_Monthly Change', 'MORTGAGE30US_Monthly Change',
       'MORTGAGE5US_Monthly Change', 'Rent Index Mon_Chg', 'NASDAQCOM Mon_Chg',
       'WILL5000 Per_Mon_Chg', 'House Completed', 'House Under Construction',
       '5yr_Maturity Per_Mon_Chg', '10yr_Maturity Per_Mon_Chg',
       '30yr_Maturity Per_Mon_Chg', 'CPI Mon_Chg',
       'House Index Monthly Change', 'Year', 'Month', 'Unemployment Rate',
       'Unemployment Rate - Black or African American',
       'Unemployment Rate - White', 'Insured Unemployment Rate',
       'Number Unemployed for 27 Weeks & Over', 'EMRATIO', 'Personal Income',
       'Real Personal Income', 'Job Openings- Total Nonfarm',
       'AverageHourlySalary-Private', 'Median Sales Price for New Houses Sold',
       'Real Private Residential Fixed Investment',
       'Total Construction Spending- Residential',
       'Monthly Supply of Houses in the United States',
       'Homeownership Rate for the United States'],
      dtype='object')

In [44]:
merged_df.to_csv("../Dataset/Final_Dataset.csv", index=False)